packages/template-retail-react-app/app/contexts/index.js

```diff
@@ -27,23 +29,157 @@ import PropTypes from 'prop-types'
  *
  * import {useCategories} from './hooks'
  *
- * const {categories, setCategories} = useCategories()
+ * const {root, findAndModifyFirst, itemsKey, fetchCategoryNode} = useCategories()
  *
  */
 export const CategoriesContext = React.createContext()
-export const CategoriesProvider = ({categories: initialCategories = {}, children}) => {
-    const [categories, setCategories] = useState(initialCategories)
+export const CategoriesProvider = ({treeRoot = {}, children}) => {
+    const itemsKey = 'categories'
+    const DEFAULT_ROOT_CATEGORY = 'root'
+    const LOCAL_STORAGE_PREFIX = `pwa-kit-cat-`
+
+    const api = useCommerceAPI()
+    const [queue, setQueue] = useState({})
+    const [root, setRoot] = useState({
+        // map over the server-provided cat
+        ...treeRoot[DEFAULT_ROOT_CATEGORY],
+        [itemsKey]: treeRoot[DEFAULT_ROOT_CATEGORY]?.[itemsKey]?.map((item) => ({
+            ...item,
+            loaded: item?.loaded ?? false
+        }))
+    })
+
+    // iterates through each deep nested object and if finds object that has prop and value specified in objToFindBy
+    // argument, it replaces the current object with replacementObj, stops recursive walk and returns the whole tree.
+    // If none is found, it returns false.
+    const findAndModifyFirst = (tree, childrenKey, objToFindBy, replacementObj) => {
+        let treeToReturn = tree
+        let findSuccess = false
+        let modifiedObj = false
+        const findKeys = Object.keys(objToFindBy)
+        findKeys.forEach((key) => {
+            isEqual(tree[key], objToFindBy[key]) ? (findSuccess = true) : (findSuccess = false)
+        })
+        if (findSuccess) {
+            for (let prop in tree) {
+                delete tree[prop]
+            }
+            for (let prop in replacementObj) {
+                tree[prop] = replacementObj[prop]
+            }
+            return tree
+        }
```

Comment:
This block is _real_ weird. It essentially transforms `tree` into `replacementObj`, while keeping the reference to `tree`. Anything that references `tree` elsewhere will be secretly changed, and I don't think that React will pick up the changes.
```suggestion
        if (findSuccess) {
            return replacementObject
        }
```
In general, state should be treated as immutable. Rather than `findAndModifyFirst(root, ...)` and changing the object referenced by `root`, you should do `const newTree = findAndModifyFirst(root, ...); setRoot(newTree)`.